### PR TITLE
Build both static and shared libraries on Unix and Windows platforms

### LINKED ISSUE #10: Build on Windows
Appears upstream supports Windows (VS 2015+ in particular). Would be good to add a build and package for it.

### PR DESCRIPTION
**Pull request by wesm**
Checklist
* [x ] Used a fork of the feedstock to propose changes
* [x] Bumped the build number (if the version is unchanged)
* [x] [Re-rendered]( https://conda-forge.org/docs/conda_smithy.html#how-to-re-render ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)

Close #10, #13

### CODE REVIEW

**Comment by wesm** [comment]
@conda-forge-admin, please rerender

**Comment by scopatz** [comment]
Thanks for putting this in @wesm! Any interest in adding yourself as a maintainer here? I think you probably care about this package more than some of the current maintainers :wink:

**Comment by wesm** [comment]
Done, I added @pitrou @xhochy and @kszucs as maintainers

**Comment by isuruf** [comment]
Does this close #10? Windows is still disabled

**Comment by scopatz** [comment]
Looks like the windows builds fail with:

```
Setting environment for using Microsoft Visual Studio 2008 x64 tools.
-- The CXX compiler identification is MSVC 15.0.30729.1
-- Check for working CXX compiler: C:/Program Files (x86)/Microsoft Visual Studio 9.0/VC/bin/amd64/cl.exe
-- Check for working CXX compiler: C:/Program Files (x86)/Microsoft Visual Studio 9.0/VC/bin/amd64/cl.exe -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
CMake Error at CMakeLists.txt:26 (message):
  you need Visual Studio 2015 or later
```

**Comment by wesm** [comment]
OK, so Python 3.5 or later I guess. I'll have to look into it later, if someone wants to push changes here please go ahead

**Comment by scopatz** [comment]
Note, that we are about to drop explicit support for python v3.5 if that helps!

**Comment by wesm** [comment]
@conda-forge-admin, please rerender

**Comment by wesm** [comment]
@conda-forge-admin, please rerender

**Comment by wesm** [comment]
@scopatz is this the right windows skip clause?

```
  skip: true  # [win and (py36 or vc<14)]
```

**Comment by scopatz** [comment]
@wesm - what are you trying to skip? That looks like it is going to skip all versions of windows, effectively

**Comment by wesm** [comment]
I tried to follow this C library pattern:

https://github.com/conda-forge/brotli-feedstock/blob/master/recipe/meta.yaml#L14

I guess that's not right. I'm copying the recipe from arrow-cpp-feedstock now (another C++11-only library)

**Comment by scopatz** [comment]
What you have now looks more correct.  I am not a deep windows build expert, so you might want to ask on gitter, but if this works and passes, I'll merge it

**Comment by wesm** [comment]
:+1:

**Comment by scopatz** [comment]
Looks like the windows build failed :cry: 

```
[ 31%] Building CXX object CMakeFiles/re2.dir/util/rune.cc.obj
rune.cc
[ 32%] Building CXX object CMakeFiles/re2.dir/util/strutil.cc.obj
strutil.cc
[ 33%] Linking CXX shared library re2.dll
NMAKE : fatal error U1077: 'C:\bld\re2_1538510221475\_build_env\Library\bin\cmake.exe' : return code '0xc0000005'
Stop.
NMAKE : fatal error U1077: '"C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\BIN\amd64\nmake.exe"' : return code '0x2'
Stop.
NMAKE : fatal error U1077: '"C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\BIN\amd64\nmake.exe"' : return code '0x2'
Stop.
Traceback (most recent call last):
  File "C:\Miniconda36-x64\Scripts\conda-build-script.py", line 10, in <module>
```

**Comment by wesm** [comment]
Hm. Guess we should try the normal MSVC build

**Comment by isuruf** [comment]
Can you try something like, https://github.com/conda-forge/snappy-feedstock/blob/4ec260701d1a8489e8a60010e215b72b7b5ec4c9/recipe/bld.bat#L3-L5 ?

**Comment by wesm** [comment]
Appears to be building now. What's this all about?

```
The system cannot find the path specified.
Traceback (most recent call last):
  File "C:\Miniconda36-x64\Scripts\conda-build-script.py", line 10, in <module>
    sys.exit(main())
  File "C:\Miniconda36-x64\lib\site-packages\conda_build\cli\main_build.py", line 438, in main
    execute(sys.argv[1:])
  File "C:\Miniconda36-x64\lib\site-packages\conda_build\cli\main_build.py", line 429, in execute
    verify=args.verify)
  File "C:\Miniconda36-x64\lib\site-packages\conda_build\api.py", line 201, in build
    notest=notest, need_source_download=need_source_download, variants=variants)
  File "C:\Miniconda36-x64\lib\site-packages\conda_build\build.py", line 2204, in build_tree
    notest=notest,
  File "C:\Miniconda36-x64\lib\site-packages\conda_build\build.py", line 1394, in build
    windows.build(m, build_file, stats=build_stats)
  File "C:\Miniconda36-x64\lib\site-packages\conda_build\windows.py", line 316, in build
    check_call_env(cmd, cwd=src_dir, stats=stats)
  File "C:\Miniconda36-x64\lib\site-packages\conda_build\utils.py", line 313, in check_call_env
    return _func_defaulting_env_to_os_environ('call', *popenargs, **kwargs)
  File "C:\Miniconda36-x64\lib\site-packages\conda_build\utils.py", line 293, in _func_defaulting_env_to_os_environ
    raise subprocess.CalledProcessError(proc.returncode, _args)
subprocess.CalledProcessError: Command '['cmd.exe', '/c', 'bld.bat']' returned non-zero exit status 1.
Command exited with code 1
```

**Comment by scopatz** [comment]
That seems odd, perhaps a conda-build issue.  Maybe bring it up on the gitter or conda-build

File: recipe/bld.bat [hunk 1]
```diff
@@ -18,3 +22,17 @@ if errorlevel 1 exit 1
...
+      -DCMAKE_BUILD_TYPE=Release ^
+      -DCMAKE_INSTALL_PREFIX=%LIBRARY_PREFIX% ^
+      -G "NMake Makefiles" ^
+      %SRC_DIR%
+if errorlevel 1 exit 1
+
+nmake
+if errorlevel 1 exit 1
+
+copy Release\re2.lib %LIBRARY_DIR%\re2_static.lib
```
Comment (isuruf) [inline recipe/bld.bat]: LIBRARY_DIR -> LIBRARY_LIB

Comment (wesm) [reply recipe/bld.bat]: fixed

File: recipe/bld.bat [hunk 2]
```diff
@@ -18,3 +22,17 @@ if errorlevel 1 exit 1
...
+cmake -DBUILD_SHARED_LIBS=OFF ^
+      -DCMAKE_BUILD_TYPE=Release ^
+      -DCMAKE_INSTALL_PREFIX=%LIBRARY_PREFIX% ^
+      -G "NMake Makefiles" ^
+      %SRC_DIR%
+if errorlevel 1 exit 1
+
+nmake
+if errorlevel 1 exit 1
+
```
Comment (scopatz) [inline recipe/bld.bat]: You are not doing an nmake install here, not sure if that is needed.

Comment (wesm) [reply recipe/bld.bat]: Can't do that because it would clobber the DDL import lib re2.lib

Comment (scopatz) [reply recipe/bld.bat]: Ahh Ok, that makes sense.  Is there a way we could turn echo on, so that we can see what is actually being run?

**Comment by wesm** [comment]
OK, adding echo shows what's wrong:

```
(C:\bld\re2_1538829964940\_build_env) C:\bld\re2_1538829964940\work\cmake-build>copy Release\re2.lib C:\bld\re2_1538829964940\_h_env\Library\lib\re2_static.lib 
The system cannot find the path specified.
(C:\bld\re2_1538829964940\_build_env) C:\bld\re2_1538829964940\work\cmake-build>if errorlevel 1 exit 1 
```

I will fix and then merge once it's working

**Comment by wesm** [comment]
@scopatz can you merge this? Thanks!

**Comment by scopatz** [comment]
Glad to see this working!

**Pull request merged**